### PR TITLE
Move auth configs into app

### PR DESCRIPTION
**Pull request by princemaple**
Refs https://github.com/firezone/product/issues/413 https://github.com/firezone/product/issues/349

image

- [x] add configs to db & schema
- [x] move oidc config parsing into start proxy
- [x] add config ui
- [x] restart oidc worker on config change
- [x] fix tests
- [x] add more tests

### CODE REVIEW

**Comment by jamilbk** [comment]
@princemaple Thanks for starting on this. Just FYI, we may not merge this for 0.5.0 just because I think some users are automating these config keys in their /etc/firezone/firezone.rb config file with Ansible or Terraform and if we moved this to the UI without some sort of API for modifying these configs, it may cause an issue.

Open to suggestions though. I was thinking after 0.5.0 is released we can tackle this config issue, v1 REST API, and Prod Docker. REST API would have a `configurations/` endpoint for addressing this.

**Comment by princemaple** [comment]
@jamilbk this will not be a breaking change.

This PR will prioritize user setting but fallback to existing configuration.

File: apps/fz_http/lib/fz_http/conf/cache.ex [hunk 1]
```diff
@@ -27,6 +27,12 @@ defmodule FzHttp.Conf.Cache do
       |> Map.delete(:id)
 
     for {k, v} <- configurations do
+      # XXX: Remove fallbacks before 1.0?
+      v =
+        with nil <- v do
+          Application.fetch_env!(:fz_http, k)
+        end
```
Comment (princemaple) [inline apps/fz_http/lib/fz_http/conf/cache.ex]: @jamilbk FYI

**Comment by jamilbk** [comment]
> @jamilbk this will not be a breaking change.
> 
> This PR will prioritize user setting but fallback to existing configuration.

Music to my ears. Well done!

File: apps/fz_http/lib/fz_http/conf/cache.ex (hunk 1 above)
Comment (jamilbk) [reply apps/fz_http/lib/fz_http/conf/cache.ex]: 👍

**Comment by princemaple** [comment]
@jamilbk give this branch a go. Haven't fixed the tests yet but the functionalities are working.

**Comment by jamilbk** [comment]
Looks great. Thanks for adopting the new config layout from @akrousset and nice using those switches.

I have some help text suggestions but this looks really awesome.

File: apps/fz_http/lib/fz_http/conf/cache.ex [hunk 2]
```diff
@@ -27,6 +29,14 @@ defmodule FzHttp.Conf.Cache do
       |> Map.delete(:id)
 
     for {k, v} <- configurations do
+      # XXX: Remove fallbacks before 1.0?
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/conf/cache.ex:32]: Yeah, this gives a graceful stepping stone. We can detect for these old config file entries (like we're doing with the old SSO config) and print an error if they're still around by next X version.

File: apps/fz_http/lib/fz_http/conf/cache.ex [hunk 3]
```diff
@@ -19,6 +19,8 @@ defmodule FzHttp.Conf.Cache do
     GenServer.start_link(__MODULE__, [])
   end
 
+  @no_fallback [:logo]
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/conf/cache.ex:22]: Cool great

File: apps/fz_http/lib/fz_http/oidc/start_proxy.ex [hunk 4]
```diff
@@ -0,0 +1,48 @@
+defmodule FzHttp.OIDC.StartProxy do
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/oidc/start_proxy.ex:1]: Nice approach in this module.

File: apps/fz_http/lib/fz_http_web/live/setting_live/security.html.heex [hunk 5]
```diff
@@ -41,4 +80,64 @@
...
+      </div>
+    </div>
+  </div>
+
+  <div class="block">
+    <strong>Auto create OIDC users</strong>
+
+    <div class="level">
+      <div class="level-left">
+        <p>Enable or disable auto creation of new users on OIDC login.</p>
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/setting_live/security.html.heex]:
```suggestion
        <p>Enable or disable auto creation of new users when logging in via OIDC for the first time.</p>
```

**Comment by princemaple** [comment]
This is ready.

**Review by jamilbk — CHANGES_REQUESTED** [review]
Initial review. Looking really good. We probably want to show a deprecation notice somewhere to start nudging users to move their config over. I think @akrousset did a hard-fail for old-style SSO config in the `omnibus/package-scripts/preinst` shell script, but maybe for now all we need is a short section in the upgrade docs at `docs/docs/administer/upgrade.md`

File: apps/fz_http/lib/fz_http/conf/cache.ex (hunk 2 above)
Comment (princemaple) [reply apps/fz_http/lib/fz_http/conf/cache.ex:32]: Was thinking something like, before 1.0, we release a 0.99 (or, say before 0.8.0, release 0.7.99). In this version, we add a temporary GenServer, that on start, reads the app configs and populates the db. And we document that:

> Before upgrading to 1.0 (or 0.8), please upgrade to 0.99 first (or 0.7.99).

**Review by jamilbk — APPROVED** [review]
lgtm

**Pull request merged**